### PR TITLE
Add clang format workflow

### PR DESCRIPTION
**Pull request by externl**

### CODE REVIEW

File: .github/workflows/clang_format.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+jobs:
+  clang-format:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Install Clang Format
+        run: |
+          # TODO: Figure out how always to install the latest version of clang-format
```
Comment (pepone) [inline .github/workflows/clang_format.yml:20]: There isn't clang-format major release that often, I think is fine if we have to manually upgrade the version from time to time when we require a feature from a more recent version.

File: .github/workflows/clang_format.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          # This LLVM script will add the relevant LLVM PPA: https://apt.llvm.org/
+          wget https://apt.llvm.org/llvm.sh -O /tmp/llvm.sh
+          chmod +x /tmp/llvm.sh
+          sudo /tmp/llvm.sh 18
+          sudo apt-get install -y clang-format-18 clang-tidy-18
```
Comment (pepone) [inline .github/workflows/clang_format.yml:25]: Why do we install `clang-tidy`?

File: cpp/src/Ice/LocatorInfo.cpp [hunk 3]
```diff
@@ -38,9 +38,8 @@ namespace
             {
                 LocatorInfo::RequestPtr request = shared_from_this();
                 _locatorInfo->getLocator()->findObjectByIdAsync(
-                    _reference->getIdentity(),
-                    [request](const optional<ObjectPrx>& object) { request->response(object); },
-                    [request](exception_ptr e) { request->exception(e); });
+                    _reference->getIdentity(), [request](const optional<ObjectPrx>& object)
```
Comment (pepone) [inline cpp/src/Ice/LocatorInfo.cpp:41]: Why this change?

File: .github/workflows/clang_format.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+          rm /tmp/llvm.sh
+
+      - name: Clang format version
+        run: clang-format-18 --version
+
+      - name: Run Clang Format
+        run: |
+          find . -name "*.h" -o -name "*.c" -o -name "*.cpp" -name "*.mm" | xargs clang-format-18 --style=file --fallback-style=none --Werror --dry-run
+          find cpp -name "*.m" | xargs clang-format-18 --style=file --fallback-style=none --Werror --dry-run
```
Comment (pepone) [inline .github/workflows/clang_format.yml:34]: How this work, does clang-format return a different status code if there are changes?

File: .github/workflows/clang_format.yml (hunk 2 above)
Comment (externl) [reply .github/workflows/clang_format.yml:25]: I'll remove it for now, would like to also run it in the future.

File: cpp/src/Ice/LocatorInfo.cpp (hunk 3 above)
Comment (externl) [reply cpp/src/Ice/LocatorInfo.cpp:41]: @bernardnormier was running version 17 when he ran it yesterday. I've updated to 18.

File: .github/workflows/clang_format.yml (hunk 4 above)
Comment (externl) [reply .github/workflows/clang_format.yml:34]: `--Werror` with `--dry-run` causes it to return an error code. Maybe we can find a nicer way to call it

File: cpp/src/Ice/LocatorInfo.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/Ice/LocatorInfo.cpp:41]: Mmm, I think the previous style was clear maybe a default value change?

File: .github/workflows/clang_format.yml (hunk 4 above)
Comment (externl) [reply .github/workflows/clang_format.yml:34]: See https://github.com/zeroc-ice/ice/actions/runs/8205903763/job/22443823637 for example

File: cpp/src/Ice/LocatorInfo.cpp (hunk 3 above)
Comment (externl) [reply cpp/src/Ice/LocatorInfo.cpp:41]: Maybe, do they have a args per line setting maybe

Comment (externl) [reply cpp/src/Ice/LocatorInfo.cpp:41]: Seems like this should be handled by `BinPackArguments`

Comment (externl) [reply cpp/src/Ice/LocatorInfo.cpp:41]: When I manually run format on save in vsocde it undoes this change. Let me look a little closer.

Comment (pepone) [reply cpp/src/Ice/LocatorInfo.cpp:41]: Right, and we set it to false, so not clear why it changed.

Comment (externl) [reply cpp/src/Ice/LocatorInfo.cpp]: ok, I've added `BinPackArguments = false`

**Comment by externl** [comment]
I pushed changes with the new `BinPackArguments` setting. Thoughts @pepone? It's the only way I could fine to get it to not pack function call arguments. But it also undoes when we have it already.

File: cpp/src/Ice/LocatorInfo.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/Ice/LocatorInfo.cpp]: I prefer this updated style.

**Review by pepone — APPROVED** [review]

**Review by bernardnormier — APPROVED** [review]
